openhtf/plugs/generic/serial_collection.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""OpenHTF plug for serial port.
+
+Allows for writing out to a serial port.
+"""
+
+import logging
+import threading
+
+try:
+  import serial
+except ImportError:
+  logging.error('Failed to import pyserial. Please install the `serial_collection_plug` extra, '
+                'e.g. via `pip install openhtf[serial_collection_plug]`.')
+  raise
+
+import openhtf
+from openhtf.util import conf
+
+
+conf.declare(
+    'serial_collection_port',
+    description='Port on which to collect serial data.',
+    default_value='/dev/ttyACM0')
+conf.declare(
+    'serial_collection_baud',
+    description='Baud rate for serial data collection.',
+    default_value=115200)
+
+
+class SerialCollectionPlug(openhtf.plugs.BasePlug):
+  """Plug that collects data from a serial port.
+
+  Spawns a thread that will open the configured serial port, continuously
+  poll the port for data, and write the data to the destination file as it is
+  received. If any serial errors are encountered during the lifetime of the
+  polling thread, data collection stops and an error message is logged.
+  Otherwise, data collection stops and the serial port is closed when
+  stop_collection() is called.
+  """
+  SERIAL_EXCEPTIONS = (serial.SerialException,
+                       serial.SerialTimeoutException,
```

Comment:
`serial.SerialTimeoutException` is a subclass of `serial.SerialException` so it's not needed here